app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_partner.rb

```diff
@@ -0,0 +1,123 @@
+
+        queries.each do |query|
+          temp_copy = ial_to_year_month_to_users.deep_dup
+
+          with_retries(
+            max_tries: 3,
+            rescue: [
+              ActiveRecord::SerializationFailure,
+              PG::ConnectionBad,
+              PG::TRSerializationFailure,
+              PG::UnableToSend,
+            ],
+            handler: proc do
+              ial_to_year_month_to_users = temp_copy
+              ActiveRecord::Base.connection.reconnect!
+            end,
+          ) do
+            Reports::BaseReport.transaction_with_timeout do
+              ActiveRecord::Base.connection.execute(query).each do |row|
+                user_id = row['user_id']
+                year_month = row['year_month']
+                auth_count = row['auth_count']
+                ial = row['ial']
+
+                ial_to_year_month_to_users[ial][year_month].add(user_id, auth_count)
+              end
+            end
+          end
+        end
+
+        rows = []
+
+        ial_to_year_month_to_users.each do |ial, year_month_to_users|
+          prev_seen_users = Set.new
+
+          year_months = year_month_to_users.keys.sort
+
+          year_months.each do |year_month|
+            year_month_users = year_month_to_users[year_month]
+
+            auth_count = year_month_users.count
+            unique_users = year_month_users.uniq.to_set
+
+            new_unique_users = unique_users - prev_seen_users
+            prev_seen_users |= unique_users
+
+            rows << {
+              key: key,
+              ial: ial,
+              year_month: year_month,
+              iaa_start_date: date_range.begin.to_s,
+              iaa_end_date: date_range.end.to_s,
+              total_auth_count: auth_count,
+              unique_users: unique_users.count,
+              new_unique_users: new_unique_users.count,
+            }
+          end
+        end
+
+        rows
+      end
+
+      # @param [Array<String>] issuers all the issuers for this iaa
+      # @param [Array<Range<Date>>] months ranges of dates by month that are included in this iaa,
+      #  the first and last may be partial months
+      # @return [Array<String>]
+      def build_queries(issuers:, months:)
+        months.map do |month_range|
+          params = {
+            range_start: month_range.begin,
+            range_end: month_range.end,
+            year_month: month_range.begin.strftime('%Y%m'),
+            issuers: issuers,
+          }.transform_values { |value| quote(value) }
+
+          format(<<~SQL, params)
+            SELECT
+              sp_return_logs.user_id
+            , %{year_month} AS year_month
+            , COUNT(sp_return_logs.id) AS auth_count
+            , sp_return_logs.ial
+            FROM sp_return_logs
+            join  service_providers on sp_return_logs.issuer = service_providers.issuer
+            WHERE
+                  sp_return_logs.profile_verified_at::date BETWEEN %{range_start} AND %{range_end}
```

Comment:
so the way this helper is supposed to work, we need calculate the age of the profile at the time of sign-in, and then group by that

So i'd expect something like (we may need to precalulate the timestamps etc


```
SELECT
# ...
SUM(
  CASE
  WHEN sp_return_logs.profile_verified_at BETWEEN %{now} AND %{one_year_ago}
  THEN 1
  ELSE 0
  END
) AS _year1,
...
SUM(
  CASE
  WHEN sp_return_logs.profile_verified_at IS UNKNOWN
  THEN 1
  ELSE 0
  END
) AS _year_unknown,
```

Alternatively we could do some sort of `GROUP BY` such as
```
GROUP BY date_extract('YEAR' from %{current_month} - profiles.verified_at)
```

Reply:
We got mixed up on aggregation levels, what you said makes sense after looking at some of the data but some other pieces are still unclear